Action that runs quarto render to check for broken links and errors

noosheenv opened the pull request:
## Internal Notes for Reviewers
I made a Github action to run `quarto render` to check for errors and broken links every time you push a commit.

The file is called `render-site.yaml` and resides in the `.github/workflows` folder.

The first few steps installs any necessary dependencies, such as Quarto and its associated extensions by using `quarto install extension`. It also tells the action to run when someone either makes a new PR or pushes something.

Then, it runs `quarto render` and checks for any immediate errors, like if you had something really off in the `_quarto.yml` file. Here's what it looks like when you try to push something with a big error:
<img width="641" alt="image" src="https://github.com/validmind/documentation/assets/143854391/2323745f-6bfd-40eb-8771-a4b90470d972">
It shows the output and also displays a message saying Quarto failed immediately.

If there are no big errors, `quarto render` will run in the background, which takes approximately 5 minutes. The workflow will then check the output for any instances of warnings or errors. If there are warnings, usually for broken links, the check will fail and display where the warning is. Here is an example:
<img width="997" alt="image" src="https://github.com/validmind/documentation/assets/143854391/bca4b587-8a1c-4bd4-a8fa-acc6003148ab">
You can find a detailed view like the screenshot above by clicking on **Details** next to the Quarto render check like this:
<img width="932" alt="image" src="https://github.com/validmind/documentation/assets/143854391/12b12e50-5221-4d9b-bf35-725a430b404e">

If there are no errors or warnings, the check will pass.

**Other changes:**
There were still some issues with broken links relating to the name of the `NLP_and_LLM` folder, so some commits fix that. Also, some release notes were referencing our hacky `NLP_and_LLM` folder, so upon fixing the folder name, the links were updated to reference `nlp_and_llm`.
<!--
PR instructions for release notes:

1. Pick at least one label:

- `internal` (skip Step 2, no release notes required)
- `highlight`
- `enhancement`
- `bug`
- `deprecation`
- `documentation`

2. In the next section, describe the changes so that an external user can understand them. Keep it simple and link to the docs with [Learn more ...](<relative-link>), if available.
-->

## External Release Notes

<!--- REPLACE THIS COMMENT WITH YOUR DESCRIPTION --->

nrichers commented on .github/workflows/render-site.yaml:
```diff
@@ -0,0 +1,49 @@
+    - name: Check for render errors 
+      run: |
+        if grep -q 'WARN:\|ERROR:' site/render_errors.log; then
+          echo "Warnings or errors detected during render"
```
```suggestion
          echo "Warnings or errors detected during Quarto render"
```

nrichers commented on .github/workflows/render-site.yaml:
```diff
@@ -0,0 +1,49 @@
+          cat site/render_errors.log
+          exit 1
+        else
+          echo "No warnings or errors detected during render"
```
```suggestion
          echo "No warnings or errors detected during Quarto render"
```

nrichers commented on .github/workflows/render-site.yaml:
```diff
@@ -0,0 +1,49 @@
+name: Check rendered site for broken links
+
+on:
+  push:
+    branches:
+      - main
```
Does your workflow run fine without these lines? We don't allow pushing directly against `main` so this should technically not do anything. If so, I'd delete.

nrichers commented on .github/workflows/render-site.yaml:
```diff
@@ -0,0 +1,49 @@
+    branches:
+      - main
+  pull_request:
+    branches:
```
I'd add a statement here that exludes draft PRs (runs the workflow only for PRs that are ready for review). You can look at https://github.com/validmind/documentation/blob/main/.github/workflows/release_notes_check.yml#L5 for an example of how that's done, the `ready_for_review` bit specifically.

nrichers commented on .github/workflows/render-site.yaml:
```diff
@@ -0,0 +1,49 @@
+    - name: Checkout repository
+      uses: actions/checkout@v2
+
+    - name: Set up Quarto
```
I'd try something like this to always grab the latest but you will need to test this actually works: 

```
- name: Set up Quarto
  run: |
    latest_version=$(curl -s https://api.github.com/repos/quarto-dev/quarto-cli/releases/latest | grep 'tag_name' | cut -d\" -f4)
    wget https://github.com/quarto-dev/quarto-cli/releases/download/${latest_version}/quarto-${latest_version}-linux-amd64.deb
    sudo dpkg -i quarto-${latest_version}-linux-amd64.deb
```

noosheenv replied:
This one was giving broken links, so I made it check the API site for the whole URL:
```
- name: Get latest Quarto release URL
      id: get-quarto-url
      run: |
        API_URL="https://api.github.com/repos/quarto-dev/quarto-cli/releases/latest"
        DOWNLOAD_URL=$(curl -s $API_URL | jq -r '.assets[] | select(.name | endswith("linux-amd64.deb")).browser_download_url')
        echo "DOWNLOAD_URL=$DOWNLOAD_URL" >> $GITHUB_ENV

    - name: Download and install Quarto
      run: |
        wget ${{ env.DOWNLOAD_URL }} -O quarto-latest-linux-amd64.deb
        sudo dpkg -i quarto-latest-linux-amd64.deb
```

nrichers commented on .github/workflows/render-site.yaml:
```diff
@@ -0,0 +1,49 @@
+        if [ -f _extensions/quarto-ext.yml ]; then
+          quarto install extension
+        fi
+# idk about this one too
```
I think this is OK but I'm not convinced it does anything, as we don't seem to have that file in our folder structure. The extensions we need are all already installed and would be included when the repo gets cloned as part of the CI check.

nrichers reviewed (commented):
Nice work, @noosheenv! Kudos on your first infrastructure PR. This looks great and also catches other errors, e.g. when an include statement is broken. 🚀 

The comments I left are mainly to tweak a few things here and there. Let me know when you've made those changes and I will retest.

noosheenv commented:
Summary of changes:
- Got rid of the on push thing, and just had the same line as the release notes check: 
```
on:
  pull_request:
    types: [opened, synchronize, labeled, unlabeled, edited, ready_for_review]
```
- Learned that Github runs workflows only if PR is able to merge first, and that's why the tests weren't running, not because of this `on: pull_request` thing
- Got Chatgpt to help me retrieve a working download link of the latest Quarto version, building off what you had (almost there but was giving error 404)
- Cleaned up comments and fixed suggested wording changes

nrichers commented on .github/workflows/render-site.yaml:
```diff
@@ -0,0 +1,44 @@
+
+on:
+  pull_request:
+    types: [opened, synchronize, labeled, unlabeled, edited, ready_for_review]
```
```suggestion
    types: [opened, synchronize, edited, ready_for_review]
```

You don't need the label function here. In fact, given the new CI check takes several minutes to run, we should minimize how often we trigger.

nrichers reviewed (commented):
@noosheenv re-reviewed and this looks good, except the new CI check runs even when the PR is in draft mode: 

<img width="1179" alt="image" src="https://github.com/validmind/documentation/assets/15148011/c5e914eb-d6d6-4b49-9a88-54952bdbc11d">

You can see in the screenshot that the release notes check was skipped — that's how it should be.

noosheenv commented:
@nrichers I seeeeee I needed an if statement under the build to skip if the PR is a draft. Should be resolved with this commit: cbabceb

nrichers reviewed (approved):
LGTM! (Looks GREAT to me.) Retested and your new GitHub workflow works as expected, including skipping when the PR is in draft mode. SHIP IT. 🚢
